KAFKA-12429: Added serdes for the default implementation of RLMM based on an internal topic as storage.

satishd opened the pull request:
KAFKA-12429:  Added serdes for the default implementation of RLMM based on an internal topic as storage. This topic will receive events of RemoteLogSegmentMetadata, RemoteLogSegmentUpdate, and RemotePartitionDeleteMetadata. These events are serialized into Kafka protocol message format.
Added tests for all the event types for that topic.

This is part of the tiered storaqe implementation [KIP-405](https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage#KIP405:KafkaTieredStorage-MessageFormat).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
KAFKA-12429: Added serdes for the default implementation of RLMM based on an internal topic as storage.

Added serdes for the default implementation of RLMM based on an internal topic as storage. This topic will receive events of RemoteLogSegmentMetadata, RemoteLogSegmentUpdate, and RemotePartitionDeleteMetadata. These events are serialized into Kafka protcocol message format.
Added tests for all the events types.

This is part of the tiered storaqge implementation KIP-405.

satishd commented:
@junrao

kowshik commented:
@satishd should this PR be reviewed after #10218 has been reviewed and merged? It seems like this PR contains some of the code thats seen in #10218, ex: `InmemoryRemoteLogMetadataManager.java`.

kieraCurtis commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManagerTest.java:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    private LogSegmentData createLogSegmentData(int segSize) throws Exception {
+        int prefix = Math.abs(new Random().nextInt());
```
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [DMI: Random object created and used only once (DMI_RANDOM_USED_ONLY_ONCE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#dmi-random-object-created-and-used-only-once-dmi-random-used-only-once).
This code creates a java.util.Random object, uses it to generate one random number, and then discards the Random object. This produces mediocre quality random numbers and is inefficient. If possible, rewrite the code so that the Random object is created once and saved, and each time a new random number is required invoke a method on the existing Random object to obtain it.
If it is important that the generated Random numbers not be guessable, you *must* not create a new Random for each random number; the values are too easily guessable. You should strongly consider using a java.security.SecureRandom instead (and avoid allocating a new SecureRandom for each random number needed).

satishd commented:
@kowshik Classes introduced in this PR are located in the new module created as part of https://github.com/apache/kafka/pull/10218. You can omit the first 2 commits and review from the third commit.

kowshik commented on storage/src/main/resources/message/RemoteLogSegmentMetadata.json:
```diff
@@ -0,0 +1,126 @@
+      "about": "End offset  of the segment."
+    },
+    {
+      "name": "BrokerId",
```
This field is not present in the KIP. The KIP contains a field called `LeaderEpoch`, which is not seen here.
Should we update the KIP suitably?

satishd replied:
Right, we will be updating the KIP with this change.

kowshik commented on storage/src/main/resources/message/RemoteLogSegmentMetadataUpdate.json:
```diff
@@ -0,0 +1,83 @@
+      ]
+    },
+    {
+      "name": "BrokerId",
```
This field is not present in the KIP. The KIP contains a field called LeaderEpoch, which is not seen here.
Should we update the KIP suitably?

satishd replied:
Right, we will be updating the KIP with this change.

kowshik commented on storage/src/main/resources/message/RemotePartitionDeleteMetadata.json:
```diff
@@ -0,0 +1,68 @@
+      ]
+    },
+    {
+      "name": "BrokerId",
```
This field is not present in the KIP. The KIP contains a field called Epoch, which is not seen here.
Should we update the KIP suitably?

satishd replied:
Right, we will be updating the KIP with this change.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serde/RemoteLogMetadataContextSerde.java:
```diff
@@ -0,0 +1,115 @@
+        rootDeserializer = (topic, data) -> deserialize(data);
+    }
+
+    private Map<Byte, RemoteLogMetadataSerdes> createInternalSerde() {
```
Can this be a static function?

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serde/RemoteLogMetadataContextSerde.java:
```diff
@@ -0,0 +1,115 @@
+    public static final byte REMOTE_LOG_SEGMENT_METADATA_UPDATE_API_KEY = (byte) new RemoteLogSegmentMetadataRecordUpdate().apiKey();
+    public static final byte REMOTE_PARTITION_DELETE_API_KEY = (byte) new RemotePartitionDeleteMetadataRecord().apiKey();
+
+    private final Map<Byte, RemoteLogMetadataSerdes> keyWithSerde;
```
Should we call this `keyToSerde`?

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serde/RemoteLogMetadataContextSerde.java:
```diff
@@ -0,0 +1,115 @@
+    private byte[] serialize(RemoteLogMetadataContext remoteLogMetadataContext) {
+        RemoteLogMetadataSerdes serDe = keyWithSerde.get(remoteLogMetadataContext.apiKey());
+        if (serDe == null) {
+            throw new IllegalArgumentException("Serializer for apikey: " + remoteLogMetadataContext.apiKey() +
```
nit: in the message s/apikey/apiKey

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serde/RemoteLogMetadataContextSerde.java:
```diff
@@ -0,0 +1,115 @@
+        return serdes;
+    }
+
+    private byte[] serialize(RemoteLogMetadataContext remoteLogMetadataContext) {
```
The `ProducerRecord` can hold a key and a value. It seems like we could store the API key in `ProducerRecord.key()` and store the serialized payload in `ProducerRecord.value()`. Why not take that route instead of serializing to a single byte array here containing (apiKey, version, payload)? 

https://kafka.apache.org/26/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html

This is of course assuming that `RemoteLogMetadataContextSerde` will only be used for serializing/deserializing events to/from a Kafka topic (as the class doc suggests).

satishd replied:
`ProducerRecord` key is generally used to route to a partition based on the key or for compaction. But we do not want to route the requests based on the apiKey here and it is not a compacted topic. 

apiKey and version are used to unmarshal the payload. IMHO, I see that as a complete structure for the record value and it should be part of the value instead of spreading in record key and value.

kowshik commented on storage/src/main/resources/message/RemoteLogSegmentMetadata.json:
```diff
@@ -0,0 +1,126 @@
+  "type": "data",
+  "name": "RemoteLogSegmentMetadataRecord",
+  "validVersions": "0",
+  "flexibleVersions": "none",
```
Could we set `flexibleVersions` to `0+` so that it means optional attributes are supported from initial version?
Same question applies to other message type definitions.

satishd replied:
Good point,  I will add that.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serde/RemoteLogSegmentMetadataSerde.java:
```diff
@@ -0,0 +1,87 @@
+    public Message serialize(RemoteLogSegmentMetadata data) {
+        RemoteLogSegmentMetadataRecord record = new RemoteLogSegmentMetadataRecord()
+                .setRemoteLogSegmentId(
+                        new RemoteLogSegmentMetadataRecord.RemoteLogSegmentIdEntry()
```
The nesting is getting deeper here, could we ease it out by extracting the code that creates a `RemoteLogSegmentIdEntry` from a `RemoteLogSegmentId` into a separate helper function? It seems like we could define it as a public static function in `RemoteLogSegmentId` and then we will be able to use it here and in `RemoteLogSegmentMetadataUpdateSerde.serialize(...)`.

satishd replied:
+1 to extract them to a different method for better readability, updated in the latest commit. 

But RemoteLogSegmentIdEntry is created respectively in RemoteLogSegmentRecord and RemoteogSegmentUpdateRecord and they can not be shared now. Is there a way to create a type and import/include it in message definitions?

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serde/RemoteLogSegmentMetadataSerde.java:
```diff
@@ -0,0 +1,87 @@
+        RemoteLogSegmentMetadata remoteLogSegmentMetadata = new RemoteLogSegmentMetadata(remoteLogSegmentId,
+                record.startOffset(), record.endOffset(), record.maxTimestampMs(), record.brokerId(),
+                record.eventTimestampMs(), record.segmentSizeInBytes(), segmentLeaderEpochs);
+        RemoteLogSegmentMetadataUpdate rlsmUpdate = new RemoteLogSegmentMetadataUpdate(remoteLogSegmentId,
```
Hmm, why do we need to create and apply an update here?

satishd replied:
We did not expose `RemoteLogSegmentMetadata` constructor to take state as it is always created with `RemoteLogSegmentState.COPY_SEGMENT_STARTED` and it should always be updated with `RemoteLogSegmentMetadataUpdate`.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serde/RemoteLogMetadataContextSerde.java:
```diff
@@ -0,0 +1,115 @@
+        return new RemoteLogMetadataContext(apiKey, version, deserializedObj);
+    }
+
+    private byte[] transformToBytes(Message message, byte apiKey, byte apiVersion) {
```
Should we rename this function as `toBytes()` and symmetrically rename the other `deserialize()` function above as `fromBytes()`?

satishd replied:
I am fine with changing `transformToBytes` to `toBytes`.  But I prefer to keep `serialize` and `deserialize` as they are the respective implementations of serializer and deserializer.

kowshik reviewed (commented):
@satishd Thanks for the PR! Few comments below on non-test code.

satishd committed:
Addressed review comments.

satishd commented:
Thanks @kowshik for the review. Addressed them with the latest [5250b79](https://github.com/apache/kafka/pull/10271/commits/5250b79cef6d21c73d2b2a12ef7cd95a00e9f034).

satishd commented:
@junrao This PR is rebased with the latest trunk and it is ready for review.

junrao commented on storage/src/main/resources/message/RemotePartitionDeleteMetadata.json:
```diff
@@ -0,0 +1,68 @@
+      "about": "Broker (controller or leader) id from which this event is created. DELETE_PARTITION_MARKED is sent by the controller. DELETE_PARTITION_STARTED and DELETE_PARTITION_FINISHED are sent by remote log metadata topic partition leader."
+    },
+    {
+      "name": "EventTimestampMs",
```
The KIP has an epoch field. Should we add it here?

satishd replied:
Initially, I thought it is good to have a leader epoch of the topic partition but I realized it is good to have broker-id instead of leader epoch for auditing purposes. I will update the KIP with these once this PR is reviewed/merged.

junrao commented on storage/src/main/resources/META-INF/services/org.apache.kafka.common.config.provider.ConfigProvider:
```diff
@@ -0,0 +1,16 @@
+ # See the License for the specific language governing permissions and
+ # limitations under the License.
+
+org.apache.kafka.common.config.provider.FileConfigProvider
```
Hmm, what is this file for?

satishd replied:
I thought this is required for code generation for message definitions in `storage/src/main/resources/message`. It looks like this is not really needed. Thanks for catching it.

junrao reviewed (commented):
@satishd : Thanks for the PR. A couple of comments below.

satishd committed:
Addressed review comments. Removed ConfigProvider file from /META-INF/services in storage module.

satishd commented:
Thanks @junrao for the review. Addressed your comments in commit [1973eda](https://github.com/apache/kafka/pull/10271/commits/1973eda32071797b8a0f090cfba9508086ae1510).

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataContextSerdes.java:
```diff
@@ -0,0 +1,116 @@
+    public static final byte REMOTE_LOG_SEGMENT_METADATA_UPDATE_API_KEY = (byte) new RemoteLogSegmentMetadataUpdateRecord().apiKey();
+    public static final byte REMOTE_PARTITION_DELETE_API_KEY = (byte) new RemotePartitionDeleteMetadataRecord().apiKey();
+
+    private final Map<Byte, RemoteLogMetadataSerdes> keyToSerdes;
```
It seems that keyToSerdes can just be a static val?

satishd replied:
Done.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerdes.java:
```diff
@@ -0,0 +1,56 @@
+     *
+     * @param metadata object to be serialized.
+     */
+    Message serialize(T metadata);
```
This is a bit weird since normally serialize() returns a ByteBuffer to be consistent with deserialize.

satishd replied:
Good point. I will refactor serialize/deserialize to return/take ByteBuffer respectively.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataContext.java:
```diff
@@ -0,0 +1,83 @@
+ *
+ * You can read more details in <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage#KIP405:KafkaTieredStorage-MessageFormat">KIP-405</a>
+ */
+public class RemoteLogMetadataContext {
```
This class is very similar to ApiMessageAndVersion. Both the metadata and the storage module need to serialize/deserialize ApiMessage. Instead of creating different variants, I am wondering if it would be better to put the common code in a new server-side module or the client module under the server package that can be shared.

satishd replied:
Sure, I can reuse `ApiMessageAndVersion`. +1 to have these in a common module. I prefer having them in a new module called `server-common`(or with better name) instead of putting it in clients module. We can slowly move server-common related code from clients module. I will raise a separate PR for this.

satishd replied:
This is slghtly different from `ApiMessageAndVersion`. This class does not have payload as the `Message` but it has `RemoteLogSegmentMetadata` or `RemoteLogSegmentMetadataUpdate` or `RemotePartitionDeleteMetadata`. This simplifies for producers/consumers of remote log metadata topic as they expect to send/receive these POJOs but not the `Message` objects. So, it is simpler for producers/consumers to have`Message` to POJO marshalling/unmarshalling in ser/des.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments below.

satishd committed:
Addressed review comments.
Refactored RemoteLogMetadataSerdes#serialize to return ByteBuffer instead of Message object

satishd commented:
Thanks @junrao for your comments. Addressed them with commit [e909c7f](https://github.com/apache/kafka/pull/10271/commits/e909c7fc481945e54b7e364d2f52c9d191f988ec).

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerdes.java:
```diff
@@ -0,0 +1,54 @@
+public interface RemoteLogMetadataSerdes<T> {
+
+    /**
+     * Returns the message serialized for the given {@code metadata} object.
```
We are no longer returning message.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataContextSerdes.java:
```diff
@@ -0,0 +1,123 @@
+        byteBuffer.put(apiVersion);
+
+        // Write the message
+        byteBuffer.put(message);
```
Hmm, this copies the payload. Could we have a standardized ApiMessage serializer that writes apiKey, apiVersion and the payload in one pass?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataContextSerdes.java:
```diff
@@ -0,0 +1,123 @@
+    private final Serializer<RemoteLogMetadataContext> rootSerializer;
+
+    public RemoteLogMetadataContextSerdes() {
+        rootSerializer = (topic, data) -> serialize(data);
```
Overall, this seems to be a convoluted way of doing serialization. (1) The logic is kind of hard to follow due to the indirections. (2) The Serde api is really designed for clients. So things like topic, configure() and close() don't really apply here.

I am wondering if we could do the following.
(a) Have a generic ApiMessageSerde that converts between a generic ApiMessageContext (with ApiMessage and version) and bytes. The serializer will write apiKey, version, payload in that order.
(b) Have a remote storage specific RemoteStorageSerde that wraps ApiMessageSerde and converts ApiMessageContext from/to RemoteStroageContext (with remote stroage specific objects). 

This way, if we have other usage of serializing ApiMessage into a log, we could reuse (a).

satishd replied:
There are only two layers currently exist for serialization. Even with ApiMessageSerde, there will be at least two layers but the ApiMessageSerde can be reused in future. 

I plan to rename RemoteLogMetadataSerde to RemoteLogMetadataTransform 
 - to avoid confusion with serde methods like seriaize/deserialize.
 - as it is really a transform of POJO to APIMessageAndVersion or vice versa.

For ApiMessageSerde, I plan to reuse ApiMessageAndVersion as you suggested earlier. This will be similar to the below code. 

```

/**
 * This abstract class provides serialization/deserialization of {@code ApiMessageAndVersion}.
 * <p></p>
 * Implementors need to extend this class and implement {@link #apiMessageFor(short)} method to return a respective
 * {@code ApiMessage} for the given {@code apiKey}. This is required to deserialize the bytes to build the respective
 * {@code ApiMessage} instance.
 */
public abstract class AbstractApiMessageAndVersionSerde implements Serde<ApiMessageAndVersion> {

    private final Serializer<ApiMessageAndVersion> serializer;
    private final Deserializer<ApiMessageAndVersion> deserializer;

    public AbstractApiMessageAndVersionSerde() {
        serializer = (topic, data) -> doSerialize(data);
        deserializer = (topic, data) -> doDeserialize(data);
    }

    private byte[] doSerialize(ApiMessageAndVersion messageAndVersion) {
        ObjectSerializationCache cache = new ObjectSerializationCache();
        short version = messageAndVersion.version();
        ApiMessage message = messageAndVersion.message();

        // Add header containing apiKey and apiVersion,
        // headerSize is 1 byte for apiKey and 1 byte for apiVersion
        int headerSize = 1 + 1;
        int messageSize = message.size(cache, version);
        ByteBufferAccessor writable = new ByteBufferAccessor(ByteBuffer.allocate(headerSize + messageSize));

        // Write apiKey and version
        writable.writeUnsignedVarint(message.apiKey());
        writable.writeUnsignedVarint(version);

        // Write the message
        message.write(writable, cache, version);

        return writable.buffer().array();
    }

    private ApiMessageAndVersion doDeserialize(byte[] data) {

        ByteBufferAccessor readable = new ByteBufferAccessor(ByteBuffer.wrap(data));

        short apiKey = (short) readable.readUnsignedVarint();
        short version = (short) readable.readUnsignedVarint();

        ApiMessage message = apiMessageFor(apiKey);
        message.read(readable, version);

        return new ApiMessageAndVersion(message, version);
    }

    /**
     * Return {@code ApiMessage} instance for the given {@code apiKey}. This is used while deserializing the bytes
     * payload into the respective {@code ApiMessage} in {@link #doDeserialize(byte[])} method.
     *
     * @param apiKey apiKey for which a {@code ApiMessage} to be created.
     */
    public abstract ApiMessage apiMessageFor(short apiKey);

    @Override
    public Serializer<ApiMessageAndVersion> serializer() {
        return serializer;
    }

    @Override
    public Deserializer<ApiMessageAndVersion> deserializer() {
        return deserializer;
    }

}

```

```

/**
 * This interface is about transforming metadata objects into the respective {@link ApiMessageAndVersion} or vice versa.
 * <p></p>
 * Those metadata objects can be {@link org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata},
 * {@link org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadataUpdate}, or {@link org.apache.kafka.server.log.remote.storage.RemotePartitionDeleteMetadata}.
 * <p>
 * @param <T> metadata type.
 *
 * @see RemoteLogSegmentMetadataTransform
 * @see RemoteLogSegmentMetadataUpdateTransform
 * @see RemotePartitionDeleteMetadataTransform
 */
public interface RemoteLogMetadataTransform<T> {

    /**
     * Transforms the given {@code metadata} object into the respective {@code ApiMessageAndVersion} object.
     *
     * @param metadata metadata object to be serialized.
     * @return transformed {@code ApiMessageAndVersion} object.
     */
    ApiMessageAndVersion toApiMessageAndVersion(T metadata);

    /**
     * Return the deserialized object for the given payload and the version.
     *
     * @param apiMessageAndVersion ApiMessageAndVersion object to be transformed.
     * @return transformed {@code T} metadata object.
     */
    T fromApiMessageAndVersion(ApiMessageAndVersion apiMessageAndVersion);

}
```

junrao replied:
Is there any benefit to use Serde? It's intended for the clients to plug-in a topic specific Serde. Here, we are creating a generic Serde for ApiMessage independent of the topics. So using Serde doesn't seem the be very natural here.

If we don't use Serde, AbstractApiMessageAndVersionSerde could just implement a serialize() and deserialize() method without topic directly, which avoids a level of indirection.

satishd replied:
Serde deserializer can be used by devs to build any tools that they need based on remote log metadata topic by setting it as value deserializer for consumer.  But they can also use byte array deserializer on consumer and invoke RemoteLogContextDeserializer for each message to convert from byte array to RemoteLogContext. I updated the PR with the suggested changes.

junrao replied:
@satishd : What you said makes sense. It's useful to be able to use the same Serde in tools or client applications. Also, it seems that it's possible to make a generic Serde for all ApiMessage since the apiKey is unique within a topic and the topic is passed into Serde.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments below.

satishd committed:
Addressed review comments.
Introduced ApiMessageAndVersionSerde and use that in RemoteLogContextSerde.

satishd commented on build.gradle:
```diff
@@ -1409,6 +1409,7 @@ project(':storage') {
   dependencies {
     implementation project(':storage:api')
     implementation project(':clients')
+    implementation project(':metadata')
```
I will remove this dependency once I move `ApiMessageAndVersion` to `clients` module. This is done in a followup PR https://github.com/apache/kafka/pull/10271.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/AbstractApiMessageAndVersionSerde.java:
```diff
@@ -0,0 +1,77 @@
+        ByteBufferAccessor writable = new ByteBufferAccessor(ByteBuffer.allocate(headerSize + messageSize));
+
+        // Write apiKey and version
+        writable.writeUnsignedVarint(message.apiKey());
```
In https://github.com/apache/kafka/blob/trunk/raft/src/main/java/org/apache/kafka/raft/metadata/MetadataRecordSerde.java#L44, the serialization adds a frame version. The intention is to use that as the header version. Currently, the only fields in the header are apikey and version. The frame version allows us to change that in the future. We need to think through whether we should adopt the same strategy here. If we do, perhaps we could just reuse MetadataRecordSerde somehow. Also, whether we should share the apiKey space with the raft messages.

satishd replied:
It adds one more indirection but it is good to have a standardized serialization mechanism for different metadata logs. 

We can introduce `AbsatrctMetadataRecordSerde` which is mostly the current `MetadataRecordSerde` excluding RAFT related `ApiMessage` generation or a given `apiKey`.  I prefer not to share the `apiKey `space across different modules but they can implement their own serde derived from `AbstractMetadataRecordSerde` by providing `ApiMessage` instances for the respective `apiKey`s.

satishd replied:
@junrao I gave a shot at introducing the mentioned changes in the earlier comment with https://github.com/apache/kafka/pull/10271/commits/e962ddf01894118655776e93907eb41a9e04a8e7.

junrao reviewed (commented):
@satishd : Another comment below.

satishd committed:
Pulled AbstractMetadataRecordSerde to have generic mechanism for metadata records serialization.
Used AbstractMetadataRecordSerde as serialization mechanism for RemoteLogContextSerde.

satishd commented on build.gradle:
```diff
@@ -1410,6 +1410,7 @@ project(':storage') {
     implementation project(':storage:api')
     implementation project(':clients')
     implementation project(':metadata')
+    implementation project(':raft')
```
I will remove this dependency once `AbstractMetadataRecordSerde` is moved to `clients` module. I plan to do that in a followup PR if we go with this approach.

satishd commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/AbstractMetadataMessageSerde.java:
```diff
@@ -19,52 +19,41 @@
  * {@code ApiMessage} instance.
  */
-public abstract class AbstractApiMessageAndVersionSerde  {
+public abstract class AbstractMetadataMessageSerde {
```
I could not find a better name here. Pl let me know if you have any.

junrao replied:
Perhaps we can use WritableApiMessageSerde and BytesApiMessageSerde for the two classes? It would be useful to write in the comment the difference between the two.

satishd replied:
`ReadableWritableApiMessageSerde` instead of `WritableApiMessageSerde`?

satishd committed:
Fixed `RemoteLogSegmentMetadataUpdateTransform` to set broker-id and updated the tests to check the same.
Minor cleanups.

junrao commented on raft/src/main/java/org/apache/kafka/raft/metadata/AbstractMetadataRecordSerde.java:
```diff
@@ -0,0 +1,82 @@
+ *
+ * This can be used as the underlying serialization mechanism for any metadata kind of log storage.
+ */
+public abstract class AbstractMetadataRecordSerde implements RecordSerde<ApiMessageAndVersion> {
```
(1) AbstractMetadataRecordSerde => WritableApiMessageSerde ?

(2) Also, could we remove the reference to raft in RecordSerde since it's no longer just for raft?

satishd replied:
(1) ReadableWritableApiMessageSerde instead of WritableApiMessageSerde?
(2) Done.

junrao replied:
(1) Perhaps we could call this one AbstractApiMessageSerde and the other one BytesApiMessageSerde?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataContext.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class RemoteLogMetadataContext {
+    private final short apiKey;
+    private final Object payload;
```
Instead of accepting arbitrary object, would it be better to have a base class like RemoteStorageMetadata for all remote log metadata? We could also use the base class to store common fields such as BrokerId and EventTimestampMs. Then, may be Serializer could just be of Serializer<RemoteStorageMetadata> instead of RemoteLogMetadataContext.

junrao commented on raft/src/main/java/org/apache/kafka/raft/metadata/AbstractMetadataRecordSerde.java:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @param apiKey apiKey for which a {@code ApiMessage} to be created.
+     */
+    public abstract ApiMessage apiMessageFor(short apiKey);
```
Another option is for all ApiMessageAndVersion that need to be stored in a log to share the same ID space. Then, we  don't need to have this method.

satishd replied:
Are you suggesting that all the metadata log related modules share the same id space? For ex: RAFT metadata, remote log metadata will share the same ID space.  IMHO, we should not go that way as it looks like an abstraction leak to me. 
 
We can have a uber serializer/deserializer at a tool level that uses the underlying serdes based on the respective metadata topic. We can introduce this serializer/deserializer in the tools module.

junrao replied:
One potential benefit of shared id space is that an ApiMessage could be used in more than one topic. There is no such use case now. I am wondering if there is such a use case in the future. If there is no such usage, we could choose not to share the id space.

satishd replied:
I guess it very unlikely to have such a use case in the future.

satishd replied:
I guess it very unlikely to have such a use case in the future.
If we encounter such a use case in the future, we can have a workaround to define a message in another topic and create a converter for one message definition to the other.

junrao commented on raft/src/main/java/org/apache/kafka/raft/metadata/AbstractMetadataRecordSerde.java:
```diff
@@ -0,0 +1,82 @@
+ * This is an implementation of {@code RecordSerde} with {@code ApiMessageAndVersion} but implementors need to implement
+ * {@link #apiMessageFor(short)} to return a {@code ApiMessage} instance for the given {@code apiKey}.
+ *
+ * This can be used as the underlying serialization mechanism for any metadata kind of log storage.
```
Could we document the serialization format in the comment?

satishd replied:
Added more details.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataContext.java:
```diff
@@ -0,0 +1,74 @@
+ * You can read more details in <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage#KIP405:KafkaTieredStorage-MessageFormat">KIP-405</a>
+ */
+public class RemoteLogMetadataContext {
+    private final short apiKey;
```
This class is now only used for holding the apiKey. We could just derive the apiKey from the specific class like RemoteLogSegmentMetadata?

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments below.

satishd committed:
Addressed review comments.
AbstractApiMessageSerde is used as the underlying serialization mechanism for remote log metadata topic.
Renamed `AbstractMetadataRecordSerde` to `AbstractApiMessageSerde`.
Renamed `RemoteLogMetadataContextSerde` to `RemoteLogMetadataSerde`.
Extracted `RemoteLogMetadata` a base class from RemoteLogSegmentMetadata etc.

junrao commented on storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadata.java:
```diff
@@ -0,0 +1,56 @@
+ * and {@link RemotePartitionDeleteMetadata}.
+ */
+@InterfaceStability.Evolving
+public class RemoteLogMetadata {
```
Should this class be abstract?

satishd commented:
Thanks @junrao for the review. Addressed them with the inline replies, and with the commit  https://github.com/apache/kafka/pull/10271/commits/265b72692f590ab5760776f2b5176066c08e959d.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerde.java:
```diff
@@ -0,0 +1,103 @@
+
+    private static final BytesApiMessageSerde BYTES_API_MESSAGE_SERDE = new BytesApiMessageSerde() {
+        @Override
+        public ApiMessage apiMessageFor(short apiKey) {
```
Could we generate this mapping using MetadataRecordTypeGenerator like what raft did?

satishd replied:
Good point. Changed remote log metadata record type as `metadata` and enabled metadata record generation.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerde.java:
```diff
@@ -0,0 +1,103 @@
+    private static final short REMOTE_LOG_SEGMENT_METADATA_UPDATE_API_KEY = new RemoteLogSegmentMetadataUpdateRecord().apiKey();
+    private static final short REMOTE_PARTITION_DELETE_API_KEY = new RemotePartitionDeleteMetadataRecord().apiKey();
+
+    private static final Map<String, Short> REMOTE_LOG_STORAGE_CLASS_TO_API_KEY = createRemoteLogStorageClassToApiKeyMap();
```
Perhaps we could just use Class as the key instead of class name?

satishd replied:
I initially had `Class` as a key but later I changed it to class name viz `String`. `Class` instances may not be equal if they are not loaded from the same classloader. String(Class name) was taken as the key to avoid such remote possible scenarios in the future.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerde.java:
```diff
@@ -0,0 +1,103 @@
+ * This class provides serialization and deserialization for {@link RemoteLogMetadata}. This is the root serde
+ * for the messages that are stored in internal remote log metadata topic.
+ */
+public class RemoteLogMetadataSerde {
```
Since we just need a single instance of this class, perhaps we could have a static public method to return the singleton and make the public constructor private?

satishd replied:
I prefer avoiding singletons if possible. I have changed this class to be open and extensible. :) This can be useful in the future if we need to extend this for tests or other purposes.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSerdeTest.java:
```diff
@@ -0,0 +1,113 @@
+
+        // Deserialize the bytes and check the RemoteLogMetadata object is as expected.
+        RemoteLogMetadataSerde deserializer = new RemoteLogMetadataSerde();
+        RemoteLogMetadata deserializedRlmContext = deserializer.deserialize(metadataBytes);
```
Should we name deserializedRlmContext differently since there is no longer a RlmContext?

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSerdeTest.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Test
+    public void testRemotePartitionDeleteMetadataSerde() {
+        // Create RemoteLogMetadataContext for RemotePartitionDeleteMetadata
```
There is no longer RemoteLogMetadataContext.

satishd replied:
sorry for running the wrong find/replace just before the commit. :(

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSerdeTest.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Test
+    public void testRemoteLogSegmentMetadataUpdateSerde() {
+        // Create RemoteLogSegmentMetadataUpdate for RemoteLogSegmentMetadataUpdate
```
Why do we need to mention RemoteLogSegmentMetadataUpdate twice?

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments below.

satishd committed:
Addressed review comments.
Changed remote log metadata message definition types as metadata and enabled metadat record generation.
`RemoteLogMetadataSerde` is made open for extension.
Cleanedup renamed/removed class references.

satishd commented:
@junrao Thanks for the review comments. Addressed them with the commit https://github.com/apache/kafka/pull/10271/commits/44cb1f374701cc6eca5d1df19dc9cd7c14497b3e

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSerdeTest.java:
```diff
@@ -0,0 +1,111 @@
+        byte[] metadataBytes = serde.serialize(remoteLogMetadata);
+
+        // Deserialize the bytes and check the RemoteLogMetadata object is as expected.
+        RemoteLogMetadataSerde deserializer = new RemoteLogMetadataSerde();
```
Do we need to instantiate the Serde a second time?

satishd replied:
Created another `RemoteLogMetadataSerde` instance to depict the real usecase of serializer and deserializer having their own instances. Added a comment on the same.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. Just one more comment below. Also, there seems to a a checkstyle error in Jenkins.

satishd committed:
Addressed review comments.
Fixed checkstyle in raft module
Added a comment in `RemoteLogMetadataSerdeTest`.

satishd commented:
Thanks @junrao for the comments. Fixed checkstyle in raft module and added a comment in `RemoteLogMetadataSerde`.

kowshik commented on raft/src/main/java/org/apache/kafka/raft/metadata/AbstractApiMessageSerde.java:
```diff
@@ -0,0 +1,93 @@
+ * This is an implementation of {@code RecordSerde} with {@link ApiMessageAndVersion} but implementors need to implement
+ * {@link #apiMessageFor(short)} to return a {@code ApiMessage} instance for the given {@code apiKey}.
+ *
+ * This can be used as the underlying serialization mechanism for any metadata kind of log storage.
```
The grammar could be improved a bit here: `... mechanism for any metadata kind of log storage.`

kowshik commented on raft/src/main/java/org/apache/kafka/raft/metadata/AbstractApiMessageSerde.java:
```diff
@@ -0,0 +1,93 @@
+ *
+ * This can be used as the underlying serialization mechanism for any metadata kind of log storage.
+ * <p></p>
+ * Serialization format for the given {@code ApiMessageAndVersion} is like below:
```
Could we drop the word `like` ?

kowshik commented on storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentMetadata.java:
```diff
@@ -270,9 +247,9 @@ public String toString() {
                ", startOffset=" + startOffset +
                ", endOffset=" + endOffset +
-               ", brokerId=" + brokerId +
+               ", brokerId=" + brokerId() +
```
Perhaps it is worth considering using `StringBuilder` here given that we are doing a number of string concatenations (~10) now.

https://docs.oracle.com/javase/7/docs/api/java/lang/StringBuilder.html

satishd replied:
afaik, javac(JDK8+) already converts that into `StringBuilder` and it is not required to replace + to `StringBuilder` concatenation code here. 
Actually, we should avoid doing that as javac has better optimizations like [JEP-280](https://openjdk.java.net/jeps/280).

kowshik replied:
Interesting, there are tons of references to `StringBuilder` in the AK code base today: https://github.com/apache/kafka/search?p=1&q=StringBuilder. Are you saying most of these are overkill?

kowshik commented on storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemotePartitionDeleteMetadata.java:
```diff
@@ -22,32 +22,30 @@
-    private final int brokerId;
 
     /**
+     * Creates an instance of this class with the given metadata.
```
We could drop this line as the intent looks obvious to me.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/BytesApiMessageSerde.java:
```diff
@@ -0,0 +1,68 @@
+import java.nio.ByteBuffer;
+
+/**
+ * This class class provides conversion of {@code ApiMessageAndVersion} to bytes and vice versa.. This can be used as serialization protocol for any
```
typo: `class class`

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerde.java:
```diff
@@ -0,0 +1,105 @@
+    private final BytesApiMessageSerde bytesApiMessageSerde;
+
+    public RemoteLogMetadataSerde() {
+        remoteLogStorageClassToApiKey = createRemoteLogStorageClassToApiKeyMap();
```
Hmm, why not instead allow the required maps (i.e. `remoteLogStorageClassToApiKey ` and `keyToTransform`) to be injected via an overloaded version of the constructor? As it stands, it feels overkill to me to ask the user to specialize this class just to inject these.

satishd replied:
Users need to implement `ApiMessage newApiMessage(short apiKey)` method if they want to reuse this class.  I kept other map initializers also left for extension.  I do not have a strong opinion on that, I am fine with your suggestion to inject them through the constructor but they still need to extend this class to implement `newApiMessage(short apiKey)`.

satishd replied:
Addressed with the latest commit https://github.com/apache/kafka/pull/10271/commits/46e0597f17e03ac751dc320e0401ae50213bf8e0

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerde.java:
```diff
@@ -0,0 +1,105 @@
+
+        return metadataTransform;
+    }
+
```
nit: can remove extra line

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerde.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    public byte[] serialize(RemoteLogMetadata remoteLogMetadata) {
+        Short apiKey = remoteLogStorageClassToApiKey.get(remoteLogMetadata.getClass().getName());
```
This 2-way map lookup feels a bit complex to me. It appears here the requirement is that you need the `apiKey` corresponding to the provided `RemoteLogMetadata`. To make it simpler, why not provide a `short apiKey()` abstract method in the `RemoteLogMetadata` base class and then ask the specializing classes to implement it? You will use then use the `remoteLogMetadata.apiKey()` method to get the apiKey here. This will avoid the need to maintain 2 maps within this class, you will only need the `remoteLogMetadataTransform` map.

satishd replied:
`apiKey` is an internal serialization detail of one of the implementations. We should not leak that to public API like `RemoteLogMetadata` and its subclasses.

kowshik replied:
I guess we can continue discussion here: https://github.com/apache/kafka/pull/10271#discussion_r622786068 ?

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/BytesApiMessageSerde.java:
```diff
@@ -0,0 +1,68 @@
+ * {@code ApiMessage} for the given {@code apiKey}. This is required to deserialize the bytes to build the respective
+ * {@code ApiMessage} instance.
+ */
+public abstract class BytesApiMessageSerde {
```
Why do we need this functionality in a separate class? Currently the only user of this class seems to be `RemoteLogMetadataSerde`. Instead if we just merge the implementation of this class into `RemoteLogMetadataSerde`, it looks simpler to me.

satishd replied:
`BytesApiMessageSerde ` is a generic implementation that can be used by others. As I said in my earlier [comment](https://github.com/apache/kafka/pull/10271#issuecomment-826991751) it will be moved to `clients` module.

kowshik replied:
@satishd I'm a little confused, where do you foresee this being reused by others i.e. other than `RemoteLogMetadataSerde`? Do you have any use cases in mind? I see that it is useful to keep it generic, but without a few clear use cases it is coming at the exchange of added complexity.

junrao replied:
I think the idea is that BytesApiMessageSerde could be reused for any future implementation that wants to serialize ApiMessage into an internal topic. AbstractApiMessageSerde allows the users to have an optimized implementation of Writable, for example, reusing a preallocated ByteBuffer (which the Raft layer uses). However, for most use cases, BytesApiMessageSerde is probably good enough and easier to use.

kowshik replied:
Sounds good.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogSegmentMetadataTransform.java:
```diff
@@ -0,0 +1,99 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class RemoteLogSegmentMetadataTransform implements RemoteLogMetadataTransform<RemoteLogSegmentMetadata> {
```
Does this class have unit test coverage?

satishd replied:
They were indirectly covered through `RemoteLogMetadataSerdeTest` but I added individual unit tests in the latest commit https://github.com/apache/kafka/pull/10271/commits/46e0597f17e03ac751dc320e0401ae50213bf8e0

kowshik replied:
Sounds good, I'll take a look.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogSegmentMetadataUpdateTransform.java:
```diff
@@ -0,0 +1,59 @@
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadataUpdate;
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentState;
+
+public class RemoteLogSegmentMetadataUpdateTransform implements RemoteLogMetadataTransform<RemoteLogSegmentMetadataUpdate> {
```
Does this class have unit test coverage?

satishd replied:
Same as https://github.com/apache/kafka/pull/10271#discussion_r623771771

kowshik replied:
Sounds good.

kowshik reviewed (commented):
@satishd Thanks for the updated PR. Few comments below.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogSegmentMetadataTransform.java:
```diff
@@ -0,0 +1,99 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class RemoteLogSegmentMetadataTransform implements RemoteLogMetadataTransform<RemoteLogSegmentMetadata> {
```
Instead of introducing this separate class, why not have `RemoteLogSegmentMetadata` implement the `RemoteLogMetadataTransform< RemoteLogSegmentMetadata>` interface directly?

satishd replied:
`RemoteLogSegmentMetadata` is a public API and we do not want to add any specific serialization implementation details into that.

kowshik replied:
@satishd interesting, so are you saying `RemoteLogSegmentMetadata` being public means it could be used outside the Broker binary? If so, where?

junrao replied:
I think what Satish means is that `RemoteLogSegmentMetadata` will be used in any remote storage plugin implementation. So, we want to be a bit careful about the public methods exposed there. Public facing methods are harder to evolve and we don't want to expose unnecessary public methods.

kowshik replied:
Sounds good.

satishd committed:
Addressed review comments.
Added RemoteLogMetadataTransformTest with unit tests for RemoteLogSegmentMetadataTransform, RemoteLogSegmentMetadataUpdateTransform, and RemotePartitionDeleteMetadataTransform.

satishd commented:
Thanks @kowshik for your comments. Addressed with replies, updated the PR with commit https://github.com/apache/kafka/pull/10271#discussion_r623771771

kowshik commented on storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadata.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    private final long eventTimestampMs;
+
+    public RemoteLogMetadata(int brokerId, long eventTimestampMs) {
```
should this be declared protected since this is an abstract class anyway?

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogSegmentMetadataTransform.java:
```diff
@@ -0,0 +1,98 @@
+                                             record.maxTimestampMs(), record.brokerId(),
+                                             record.eventTimestampMs(), record.segmentSizeInBytes(),
+                                             segmentLeaderEpochs);
+        RemoteLogSegmentMetadataUpdate rlsmUpdate
```
Maybe I'm missing something, but there seems to be a `RemoteLogSegmentMetadata` constructor overload that also allows you to pass the `RemoteLogSegmentState` as one of the parameters: https://github.com/apache/kafka/blob/trunk/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentMetadata.java#L99

If we use the overloaded version, do we still need to apply the `RemoteLogSegmentMetadataUpdate` here?

satishd replied:
That constructor is kept private and wanted to control what fields can be changed on the existing instance with `RemoteLogSegmentMetadataUpdate`.

kowshik replied:
Makes sense.

kowshik reviewed (commented):
@satishd: Thanks for the updated PR! LGTM. Just couple more small final comments below.

satishd committed:
Addressed reveiw comments.
Changed RemoteLogMetadata constructor access from public to protected as it is an abstract class.

junrao reviewed (approved):
@satishd : Thanks for the updated PR. LGTM

dengziming commented on raft/src/main/java/org/apache/kafka/raft/metadata/AbstractApiMessageSerde.java:
```diff
@@ -0,0 +1,93 @@
+     *
+     * @param apiKey apiKey for which a {@code ApiMessage} to be created.
+     */
+    public abstract ApiMessage apiMessageFor(short apiKey);
```
Hello @satishd , why should we make this abstract? I find that the two implementations of this method are all `MetadataRecordType.fromId(apiKey).newMetadataRecord()`.

satishd replied:
@dengziming Below are the two implementations of `AbstractApiMessageSerde` and they do not share same records types.
- [BytesApiMessageSerde](https://github.com/apache/kafka/blob/trunk/server-common/src/main/java/org/apache/kafka/server/common/serialization/BytesApiMessageSerde.java#L41rl)
```BytesApiMessageSerde.this.apiMessageFor(apiKey);```

- [MetadataRecordSerde](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/MetadataRecordSerde.java#L27)
```MetadataRecordType.fromId(apiKey).newMetadataRecord();```
